plugins/app-backend/src/service/router.ts

```diff
@@ -123,26 +134,162 @@ export async function createRouter(
+        // and continue as unauthenticated. If the user is logged in they will
+        // immediately be redirected back to the protected app via the POST.
+        await httpAuth.issueUserCookie(res, {
+          credentials: await auth.getNoneCredentials(),
+        });
+        next();
+      }
+    });
+
+    publicRouter.post(
+      '*',
+      express.urlencoded({ extended: true }),
+      async (req, res, next) => {
+        if (req.body.type === 'sign-in') {
+          const credentials = await auth.authenticate(req.body.token);
+
+          if (!auth.isPrincipal(credentials, 'user')) {
+            throw new AuthenticationError('Invalid token, not a user');
+          }
+
+          await httpAuth.issueUserCookie(res, {
+            credentials,
+          });
+
+          // Resume as if it was a GET request towards the outer protected router, serving index.html
+          req.method = 'GET';
+          next('router');
+        } else {
+          throw new Error('Invalid POST request to /');
+        }
+      },
+    );
+
+    publicRouter.use(
+      await createEntryPointRouter({
+        appMode: 'public',
+        logger: logger.child({ entry: 'public' }),
+        rootDir: publicDistDir,
+        assetStore: assetStore?.withNamespace('public'),
+        appConfigs, // TODO(Rugvip): We should not be including the full config here
+      }),
+    );
+
+    router.use(publicRouter);
+  }
+
+  router.use(
+    await createEntryPointRouter({
+      appMode: enablePublicEntryPoint ? 'protected' : 'public',
+      logger: logger.child({ entry: 'main' }),
+      rootDir: appDistDir,
+      assetStore,
+      staticFallbackHandler,
+      appConfigs,
+    }),
+  );
+
+  return router;
+}
+
+async function injectAppMode(options: {
+  appMode: 'public' | 'protected';
+  rootDir: string;
+}) {
+  const { appMode, rootDir } = options;
+  const content = await fs.readFile(resolvePath(rootDir, 'index.html'), 'utf8');
+
+  const metaTag = `<meta name="backstage-app-mode" content="${appMode}">`;
+
+  let newContent;
+  if (content.includes('backstage-app-mode')) {
+    newContent = content.replace(
+      /<meta name="backstage-app-mode" content="[^"]+">/,
+      metaTag,
+    );
+  } else {
+    newContent = content.replace(/<head>/, `<head>${metaTag}`);
+  }
+
+  await fs.writeFile(resolvePath(rootDir, 'index.html'), newContent, 'utf8');
+}
+
+async function createEntryPointRouter({
+  logger,
+  rootDir,
+  assetStore,
+  staticFallbackHandler,
+  appMode,
+  appConfigs,
+}: {
+  logger: Logger;
+  rootDir: string;
+  assetStore?: StaticAssetsStore;
+  staticFallbackHandler?: express.Handler;
+  appMode: 'public' | 'protected';
+  appConfigs?: AppConfig[];
+}) {
+  const staticDir = resolvePath(rootDir, 'static');
+
+  const injectedConfigPath =
+    appConfigs && (await injectConfig({ appConfigs, logger, staticDir }));
+
+  await injectAppMode({ appMode, rootDir });
```

Comment:
Just noting that our `disableConfigInjection` was there to cover readonly filesystems, and now this one also does writes - do we have to honor that flag for this purpose too?

Also, is it really not possible to do "live" injection of stuff in our app backend serving code? Even with like a middleware, that mutates the body before getting sent?

Reply:
We had a live one but switched to this for simplicity, since that one needs to handle a couple of different cases. We could tweak the injection to only happen if it's not needed, but tbh I prefer the approach of adding it on the fly. Gonna leave that for a followup though